File: Horizon-CopilotKit/packages/runtime/src/service-adapters/deepseek/deepseek-adapter.ts
/**
 * Copilot Runtime adapter for DeepSeek.
 *
 * ## Example
 *
 * ```ts
 * import { CopilotRuntime, DeepSeekAdapter } from "@copilotkit/runtime";
 * import OpenAI from "openai";
 *
 * const copilotKit = new CopilotRuntime();
 *
 * const deepseek = new OpenAI({
 *   apiKey: "<your-deepseek-api-key>",
 *   baseURL: "https://api.deepseek.com/v1",
 * });
 *
 * return new DeepSeekAdapter({ openai: deepseek });
 * ```
 *
 * ## Available Models
 *
 * DeepSeek supports the following models:
 * - deepseek-chat (default): DeepSeek's flagship model optimized for chat
 * - deepseek-coder: Specialized for code generation and understanding
 * - deepseek-reasoner: Enhanced reasoning capabilities
 */
import OpenAI from "openai";
import {
  CopilotServiceAdapter,
  CopilotRuntimeChatCompletionRequest,
  CopilotRuntimeChatCompletionResponse,
} from "../service-adapter";
import {
  convertActionInputToOpenAITool,
  convertMessageToOpenAIMessage,
  limitMessagesToTokenCount,
} from "../openai/utils";
import { randomUUID } from "@copilotkit/shared";

const DEFAULT_MODEL = "deepseek-chat";
const DEEPSEEK_BASE_URL = "https://api.deepseek.com/v1";

export interface DeepSeekAdapterParams {
  /**
   * An OpenAI-compatible instance configured for DeepSeek.
   * If not provided, a new instance will be created with DeepSeek's base URL.
   */
  openai?: OpenAI;

  /**
   * DeepSeek API key. Required if openai instance is not provided.
   */
  apiKey?: string;

  /**
   * The DeepSeek model to use.
   * Available models: deepseek-chat, deepseek-coder, deepseek-reasoner
   *
   * @default "deepseek-chat"
   */
  model?: string;

  /**
   * Whether to disable parallel tool calls.
   * DeepSeek supports parallel tool calls for efficiency.
   *
   * @default false
   */
  disableParallelToolCalls?: boolean;

  /**
   * Custom base URL for DeepSeek API.
   * 
   * @default "https://api.deepseek.com/v1"
   */
  baseURL?: string;

  /**
   * Additional headers to send with requests.
   */
  headers?: Record<string, string>;
}

export class DeepSeekAdapter implements CopilotServiceAdapter {
  private model: string = DEFAULT_MODEL;
  private disableParallelToolCalls: boolean = false;
  private _openai: OpenAI;

  public get openai(): OpenAI {
    return this._openai;
  }

  constructor(params?: DeepSeekAdapterParams) {
    if (params?.openai) {
      this._openai = params.openai;
    } else {
      if (!params?.apiKey) {
        throw new Error("DeepSeek API key is required when openai instance is not provided");
      }
      
      this._openai = new OpenAI({
        apiKey: params.apiKey,
        baseURL: params?.baseURL || DEEPSEEK_BASE_URL,
        defaultHeaders: {
          "User-Agent": "CopilotKit-DeepSeek-Adapter",
          ...params?.headers,
        },
      });
    }

    if (params?.model) {
      this.model = params.model;
    }
    this.disableParallelToolCalls = params?.disableParallelToolCalls || false;
  }

  async process(
    request: CopilotRuntimeChatCompletionRequest,
  ): Promise<CopilotRuntimeChatCompletionResponse> {
    const {
      threadId: threadIdFromRequest,
      model = this.model,
      messages,
      actions,
      eventSource,
      forwardedParameters,
    } = request;
    console.log("🔄 [DeepSeek] Processing request:", {
      threadId: threadIdFromRequest,
      model,
      messagesCount: messages.length,
      actionsCount: actions.length,
      timestamp: new Date().toISOString()
    });
    const tools = actions.map(convertActionInputToOpenAITool);
    const threadId = threadIdFromRequest ?? randomUUID();

    // ALLOWLIST APPROACH: Only include tool_result messages that correspond to valid tool_calls
    // Step 1: Extract valid tool_call IDs
    const validToolUseIds = new Set<string>();

    for (const message of messages) {
      if (message.isActionExecutionMessage()) {
        validToolUseIds.add(message.id);
      }
    }

    // Step 2: Filter messages, keeping only those with valid tool_call IDs
    const filteredMessages = messages.filter((message) => {
      if (message.isResultMessage()) {
        // Skip if there's no corresponding tool_call
        if (!validToolUseIds.has(message.actionExecutionId)) {
          return false;
        }

        // Remove this ID from valid IDs so we don't process duplicates
        validToolUseIds.delete(message.actionExecutionId);
        return true;
      }

      // Keep all non-tool-result messages
      return true;
    });

    let openaiMessages = filteredMessages.map((m) =>
      convertMessageToOpenAIMessage(m, { 
        // DeepSeek prefers 'system' role for system messages
        keepSystemRole: true 
      }),
    );
    
    // 🔧 DeepSeek 兼容性修复：将不支持的 'developer' 角色转换为 'system'
    openaiMessages = openaiMessages.map((message) => {
      if (message && typeof message === 'object' && 'role' in message && message.role === 'developer') {
        console.log('🔄 [DeepSeek] Converting developer role to system role');
        return {
          ...message,
          role: 'system' as const,
        };
      }
      return message;
    });
    
    openaiMessages = limitMessagesToTokenCount(openaiMessages, tools, model);

    let toolChoice: any = forwardedParameters?.toolChoice;
    if (forwardedParameters?.toolChoice === "function") {
      toolChoice = {
        type: "function",
        function: { name: forwardedParameters.toolChoiceFunctionName },
      };
    }

    console.log("📤 [DeepSeek] Sending request to API:", {
      model,
      messagesCount: openaiMessages.length,
      toolsCount: tools.length,
      messages: openaiMessages.map(m => ({ role: m.role, content: typeof m.content === 'string' ? m.content.substring(0, 100) + '...' : typeof m.content }))
    });

    try {
      const requestPayload = {
        model: model,
        stream: true as const,
        messages: openaiMessages,
        ...(tools.length > 0 && { tools }),
        ...(forwardedParameters?.maxTokens && { max_tokens: forwardedParameters.maxTokens }),
        ...(forwardedParameters?.stop && { stop: forwardedParameters.stop }),
        ...(toolChoice && { tool_choice: toolChoice }),
        ...(this.disableParallelToolCalls && { parallel_tool_calls: false }),
        ...(forwardedParameters?.temperature && { 
          temperature: Math.max(0.1, Math.min(2.0, forwardedParameters.temperature)) // DeepSeek temperature range
        }),
      };

      console.log("📤 [DeepSeek] API Request payload:", JSON.stringify(requestPayload, null, 2));

      const stream = this.openai.beta.chat.completions.stream(requestPayload);

      console.log("🔄 [DeepSeek] Stream created successfully, starting to process...");
      console.log("🔄 [DeepSeek] Stream object:", typeof stream, stream.constructor.name);

      eventSource.stream(async (eventStream$) => {
        let mode: "function" | "message" | null = null;
        let currentMessageId: string = "";
        let currentToolCallId: string = "";
        let currentActionName: string = ""; // 跟踪当前执行的 Action 名称

        try {
          console.log("🔄 [DeepSeek] Starting stream iteration...");
          let chunkCount = 0;
          let lastChunkTime = Date.now();
          
          for await (const chunk of stream) {
            lastChunkTime = Date.now();
            chunkCount++;
            
            console.log(`📦 [DeepSeek] Received chunk #${chunkCount}:`, { 
              choicesLength: chunk.choices.length,
              finishReason: chunk.choices[0]?.finish_reason,
              hasToolCall: !!chunk.choices[0]?.delta.tool_calls?.[0],
              hasContent: !!chunk.choices[0]?.delta.content,
            });

            if (chunk.choices.length === 0) {
              continue;
            }

            const toolCall = chunk.choices[0].delta.tool_calls?.[0];
            const content = chunk.choices[0].delta.content;
            const finishReason = chunk.choices[0].finish_reason;

            // 🔧 检查是否应该结束流
            if (finishReason) {
              console.log(`🏁 [DeepSeek] Finish reason detected: ${finishReason}, will end after processing this chunk`);
            }

            // 🔧 关键修复：使用 OpenAI 适配器的模式切换逻辑
            // When switching from message to function or vice versa, send the respective end event.
            if (mode === "message" && toolCall?.id) {
              console.log("🔧 [DeepSeek] Switching from message to function mode");
              mode = null;
              eventStream$.sendTextMessageEnd({ messageId: currentMessageId });
            } else if (mode === "function" && (toolCall === undefined || toolCall?.id)) {
              console.log("🔧 [DeepSeek] Switching from function to message mode");
              mode = null;
              eventStream$.sendActionExecutionEnd({ actionExecutionId: currentToolCallId });
            }

            // If we send a new message type, send the appropriate start event.
            if (mode === null) {
              if (toolCall?.id) {
                console.log("🚀 [DeepSeek] Starting function mode");
                mode = "function";
                currentToolCallId = toolCall.id;
                currentActionName = toolCall.function?.name || "";
                eventStream$.sendActionExecutionStart({
                  actionExecutionId: currentToolCallId,
                  parentMessageId: chunk.id,
                  actionName: currentActionName,
                });
              } else if (content) {
                console.log("💬 [DeepSeek] Starting message mode");
                mode = "message";
                currentMessageId = chunk.id || randomUUID();
                eventStream$.sendTextMessageStart({ messageId: currentMessageId });
              }
            }

            // send the content events
            if (mode === "message" && content) {
              console.log("💬 [DeepSeek] Sending text content:", content);
              eventStream$.sendTextMessageContent({
                messageId: currentMessageId,
                content: content,
              });
            } else if (mode === "function" && toolCall?.function?.arguments) {
              console.log("📝 [DeepSeek] Sending function arguments:", toolCall.function.arguments);
              eventStream$.sendActionExecutionArgs({
                actionExecutionId: currentToolCallId,
                args: toolCall.function.arguments,
              });
            }

            // 🔧 强制在 finishReason 后结束流
            if (finishReason) {
              console.log(`🔚 [DeepSeek] Breaking loop due to finish reason: ${finishReason}`);
              break;
            }
          }

          // 🔧 关键修复：确保流正确结束 - 来自 OpenAI 适配器  
          console.log(`🏁 [DeepSeek] Stream loop ended after ${chunkCount} chunks, sending final events`);
          
          // send the end events
          if (mode === "message") {
            console.log("💬 [DeepSeek] Ending final text message");
            eventStream$.sendTextMessageEnd({ messageId: currentMessageId });
          } else if (mode === "function") {
            console.log("🔧 [DeepSeek] Ending final function execution");
            eventStream$.sendActionExecutionEnd({ actionExecutionId: currentToolCallId });
          }
        } catch (error) {
          console.error("❌ [DeepSeek] streaming error:", error);
          if (mode === "message") {
            console.log("💬 [DeepSeek] Error cleanup: ending text message");
            eventStream$.sendTextMessageEnd({
              messageId: currentMessageId,
            });
          } else if (mode === "function" && currentToolCallId) {
            console.log("🔧 [DeepSeek] Error cleanup: ending function execution");
            eventStream$.sendActionExecutionEnd({
              actionExecutionId: currentToolCallId,
            });
          }
          throw error;
        }

        // 🔧 关键修复：明确完成事件流 - 来自 OpenAI 适配器
        console.log("🎉 [DeepSeek] Completing event stream");
        eventStream$.complete();
      });

      return { threadId };
    } catch (error) {
      console.error("DeepSeek API error:", error);
      throw new Error(`DeepSeek API request failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }
} 

File: Horizon-CopilotKit/packages/runtime-python/copilotkit_runtime/adapters/deepseek_adapter.py
"""
CopilotKit 运行时的 DeepSeek 适配器

DeepSeek 是一家专注于 AI 研究的公司，提供强大的大语言模型服务。
本适配器为 CopilotKit 提供对 DeepSeek AI 模型的完整支持。

## 使用示例

```python
from copilotkit_runtime import CopilotRuntime, DeepSeekAdapter
from openai import AsyncOpenAI

# 创建运行时实例
copilot_kit = CopilotRuntime()

# 方法 1: 直接使用 API Key
adapter = DeepSeekAdapter(
    api_key="your-deepseek-api-key",
    model="deepseek-chat"
)

# 方法 2: 使用预配置的 OpenAI 客户端
deepseek = AsyncOpenAI(
    api_key="your-deepseek-api-key",
    base_url="https://api.deepseek.com/v1",
)
adapter = DeepSeekAdapter(openai=deepseek)
```

## 支持的模型

DeepSeek 支持以下模型：
- deepseek-chat (默认): DeepSeek 的旗舰对话模型，平衡性能和质量
- deepseek-coder: 专门针对代码生成和理解优化的模型
- deepseek-reasoner: 增强推理能力的模型，适合复杂问题解决

## 特性

- ✅ 完整兼容 CopilotKit Runtime API
- ✅ 支持流式响应和实时文本生成
- ✅ 支持并行和串行工具调用
- ✅ 自动温度范围限制（0.1-2.0）
- ✅ 角色转换（developer -> system）
- ✅ 完善的错误处理和日志记录
"""

import asyncio
import json
import logging
from typing import Optional, Dict, Any, AsyncGenerator
from uuid import uuid4

from openai import AsyncOpenAI
from openai.types.chat import ChatCompletionChunk

from .base import CopilotServiceAdapter
from ..types import (
    CopilotRuntimeChatCompletionRequest,
    CopilotRuntimeChatCompletionResponse,
)
from ..utils import (
    convert_action_to_openai_tool,
    convert_message_to_openai_message,
    limit_messages_to_token_count,
)
from ..events import EventSource

logger = logging.getLogger(__name__)

DEFAULT_MODEL = "deepseek-chat"
DEEPSEEK_BASE_URL = "https://api.deepseek.com/v1"


class DeepSeekAdapter(CopilotServiceAdapter):
    """
    CopilotKit 运行时的 DeepSeek 适配器
    
    提供与 DeepSeek AI 模型的集成，包括 deepseek-chat、
    deepseek-coder 和 deepseek-reasoner 模型。
    
    此适配器负责：
    - 将 CopilotKit 请求转换为 DeepSeek API 调用
    - 处理流式响应和事件流
    - 管理工具调用和参数传递
    - 处理错误和异常情况
    - 提供完整的日志记录
    
    兼容性：
    - 与 TypeScript 版本的 DeepSeekAdapter 完全兼容
    - 支持所有 OpenAI 兼容的参数和配置
    - 自动处理 DeepSeek 特定的限制和要求
    """
    
    def __init__(
        self,
        openai: Optional[AsyncOpenAI] = None,
        api_key: Optional[str] = None,
        model: str = DEFAULT_MODEL,
        disable_parallel_tool_calls: bool = False,
        base_url: str = DEEPSEEK_BASE_URL,
        headers: Optional[Dict[str, str]] = None,
    ):
        """
        初始化 DeepSeek 适配器
        
        参数:
            openai: 预配置的 AsyncOpenAI 实例，用于 DeepSeek API 调用
            api_key: DeepSeek API 密钥（如果未提供 openai 实例则必需）
            model: 要使用的 DeepSeek 模型
                   - deepseek-chat: 通用对话模型（默认）
                   - deepseek-coder: 代码专用模型
                   - deepseek-reasoner: 推理增强模型
            disable_parallel_tool_calls: 是否禁用并行工具调用
            base_url: 自定义 DeepSeek API 基础 URL
            headers: 发送请求时的额外头部信息
        
        异常:
            ValueError: 当未提供 openai 实例且缺少 api_key 时抛出
        
        使用示例:
            # 使用 API Key 初始化
            adapter = DeepSeekAdapter(api_key="sk-xxx", model="deepseek-chat")
            
            # 使用预配置客户端初始化
            client = AsyncOpenAI(api_key="sk-xxx", base_url="https://api.deepseek.com/v1")
            adapter = DeepSeekAdapter(openai=client)
        """
        if openai:
            self._openai = openai
        else:
            if not api_key:
                raise ValueError("DeepSeek API key is required when openai instance is not provided")
            
            default_headers = {
                "User-Agent": "CopilotKit-DeepSeek-Adapter",
            }
            if headers:
                default_headers.update(headers)
                
            self._openai = AsyncOpenAI(
                api_key=api_key,
                base_url=base_url,
                default_headers=default_headers,
            )
        
        self.model = model
        self.disable_parallel_tool_calls = disable_parallel_tool_calls
        
    @property
    def openai(self) -> AsyncOpenAI:
        """Get the OpenAI client instance."""
        return self._openai
    
    async def process(
        self, request: CopilotRuntimeChatCompletionRequest
    ) -> CopilotRuntimeChatCompletionResponse:
        """Process chat completion request using DeepSeek."""
        thread_id = request.thread_id or str(uuid4())
        model = request.model or self.model
        messages = request.messages
        actions = request.actions
        event_source = request.event_source
        forwarded_parameters = request.forwarded_parameters or {}
        
        logger.info(
            "🔄 [DeepSeek] Processing request: thread_id=%s, model=%s, messages=%d, actions=%d",
            thread_id, model, len(messages), len(actions)
        )
        
        # Convert actions to OpenAI tools
        tools = [convert_action_to_openai_tool(action) for action in actions]
        
        # Filter messages using allowlist approach
        valid_tool_use_ids = set()
        
        # Extract valid tool_call IDs
        for message in messages:
            if message.is_action_execution_message():
                valid_tool_use_ids.add(message.id)
        
        # Filter messages, keeping only those with valid tool_call IDs
        filtered_messages = []
        for message in messages:
            if message.is_result_message():
                # Skip if there's no corresponding tool_call
                if message.action_execution_id not in valid_tool_use_ids:
                    continue
                # Remove this ID from valid IDs so we don't process duplicates
                valid_tool_use_ids.discard(message.action_execution_id)
            filtered_messages.append(message)
        
        # Convert to OpenAI messages
        openai_messages = [
            convert_message_to_openai_message(m, keep_system_role=True)
            for m in filtered_messages
        ]
        
        # DeepSeek compatibility fix: convert unsupported 'developer' role to 'system'
        for message in openai_messages:
            if isinstance(message, dict) and message.get("role") == "developer":
                logger.info("🔄 [DeepSeek] Converting developer role to system role")
                message["role"] = "system"
        
        # Limit messages to token count
        openai_messages = limit_messages_to_token_count(openai_messages, tools, model)
        
        # Prepare tool choice
        tool_choice = forwarded_parameters.get("tool_choice")
        if tool_choice == "function":
            tool_choice = {
                "type": "function",
                "function": {"name": forwarded_parameters.get("tool_choice_function_name")},
            }
        
        logger.info(
            "📤 [DeepSeek] Sending request to API: model=%s, messages=%d, tools=%d",
            model, len(openai_messages), len(tools)
        )
        
        try:
            # Prepare request payload
            request_payload = {
                "model": model,
                "stream": True,
                "messages": openai_messages,
            }
            
            if tools:
                request_payload["tools"] = tools
            
            if forwarded_parameters.get("max_tokens"):
                request_payload["max_tokens"] = forwarded_parameters["max_tokens"]
                
            if forwarded_parameters.get("stop"):
                request_payload["stop"] = forwarded_parameters["stop"]
                
            if tool_choice:
                request_payload["tool_choice"] = tool_choice
                
            if self.disable_parallel_tool_calls:
                request_payload["parallel_tool_calls"] = False
                
            if forwarded_parameters.get("temperature"):
                # DeepSeek temperature range: 0.1-2.0
                temperature = max(0.1, min(2.0, forwarded_parameters["temperature"]))
                request_payload["temperature"] = temperature
            
            logger.debug("📤 [DeepSeek] API Request payload: %s", json.dumps(request_payload, indent=2))
            
            # Create stream
            stream = await self._openai.chat.completions.create(**request_payload)
            
            logger.info("🔄 [DeepSeek] Stream created successfully, starting to process...")
            
            # Process stream
            await self._process_stream(stream, event_source, thread_id)
            
            return CopilotRuntimeChatCompletionResponse(thread_id=thread_id)
            
        except Exception as error:
            logger.error("❌ [DeepSeek] API error: %s", error)
            raise RuntimeError(f"DeepSeek API request failed: {error}")
    
    async def _process_stream(
        self,
        stream: AsyncGenerator[ChatCompletionChunk, None],
        event_source: EventSource,
        thread_id: str,
    ) -> None:
        """Process the streaming response from DeepSeek."""
        
        async def stream_handler(event_stream):
            mode = None  # "function" | "message" | None
            current_message_id = ""
            current_tool_call_id = ""
            current_action_name = ""
            
            try:
                logger.info("🔄 [DeepSeek] Starting stream iteration...")
                chunk_count = 0
                
                async for chunk in stream:
                    chunk_count += 1
                    
                    logger.debug(
                        "📦 [DeepSeek] Received chunk #%d: choices=%d, finish_reason=%s",
                        chunk_count,
                        len(chunk.choices),
                        chunk.choices[0].finish_reason if chunk.choices else None,
                    )
                    
                    if not chunk.choices:
                        continue
                    
                    choice = chunk.choices[0]
                    tool_call = choice.delta.tool_calls[0] if choice.delta.tool_calls else None
                    content = choice.delta.content
                    finish_reason = choice.finish_reason
                    
                    if finish_reason:
                        logger.info("🏁 [DeepSeek] Finish reason detected: %s", finish_reason)
                    
                    # Mode switching logic (from OpenAI adapter)
                    if mode == "message" and tool_call and tool_call.id:
                        logger.info("🔧 [DeepSeek] Switching from message to function mode")
                        mode = None
                        await event_stream.send_text_message_end(message_id=current_message_id)
                    elif mode == "function" and (not tool_call or tool_call.id):
                        logger.info("🔧 [DeepSeek] Switching from function to message mode")
                        mode = None
                        await event_stream.send_action_execution_end(
                            action_execution_id=current_tool_call_id
                        )
                    
                    # Start new mode if needed
                    if mode is None:
                        if tool_call and tool_call.id:
                            logger.info("🚀 [DeepSeek] Starting function mode")
                            mode = "function"
                            current_tool_call_id = tool_call.id
                            current_action_name = tool_call.function.name if tool_call.function else ""
                            await event_stream.send_action_execution_start(
                                action_execution_id=current_tool_call_id,
                                parent_message_id=chunk.id,
                                action_name=current_action_name,
                            )
                        elif content:
                            logger.info("💬 [DeepSeek] Starting message mode")
                            mode = "message"
                            current_message_id = chunk.id or str(uuid4())
                            await event_stream.send_text_message_start(message_id=current_message_id)
                    
                    # Send content events
                    if mode == "message" and content:
                        logger.debug("💬 [DeepSeek] Sending text content: %s", content)
                        await event_stream.send_text_message_content(
                            message_id=current_message_id,
                            content=content,
                        )
                    elif mode == "function" and tool_call and tool_call.function and tool_call.function.arguments:
                        logger.debug("📝 [DeepSeek] Sending function arguments: %s", tool_call.function.arguments)
                        await event_stream.send_action_execution_args(
                            action_execution_id=current_tool_call_id,
                            args=tool_call.function.arguments,
                        )
                    
                    # Break on finish reason
                    if finish_reason:
                        logger.info("🔚 [DeepSeek] Breaking loop due to finish reason: %s", finish_reason)
                        break
                
                # Send final end events
                logger.info("🏁 [DeepSeek] Stream loop ended after %d chunks, sending final events", chunk_count)
                
                if mode == "message":
                    logger.info("💬 [DeepSeek] Ending final text message")
                    await event_stream.send_text_message_end(message_id=current_message_id)
                elif mode == "function":
                    logger.info("🔧 [DeepSeek] Ending final function execution")
                    await event_stream.send_action_execution_end(
                        action_execution_id=current_tool_call_id
                    )
                    
            except Exception as error:
                logger.error("❌ [DeepSeek] streaming error: %s", error)
                
                # Error cleanup
                if mode == "message":
                    logger.info("💬 [DeepSeek] Error cleanup: ending text message")
                    await event_stream.send_text_message_end(message_id=current_message_id)
                elif mode == "function" and current_tool_call_id:
                    logger.info("🔧 [DeepSeek] Error cleanup: ending function execution")
                    await event_stream.send_action_execution_end(
                        action_execution_id=current_tool_call_id
                    )
                raise error
            
            # Complete event stream
            logger.info("🎉 [DeepSeek] Completing event stream")
            await event_stream.complete()
        
        await event_source.stream(stream_handler) 